Differential expression RNA-seq Revamp (#183)
AlexsLemonade/refinebio-examples: 03-rnaseq/differential-expression_rnaseq_01.Rmd

---
title: "Differential Expression - RNA-seq"
author: "CCDL for ALSF"
date: "`r format(Sys.time(), '%B %Y')`"
output:   
  html_notebook:
    toc: true
    toc_float: true
    number_sections: true
---

# Purpose of the analysis:

> <!--WHAT DOES THIS ANALYSIS DO AND WHEN WOULD YOU USE IT?-->

Author: We introduce these to help reviewing because it's otherwise hard to see what information is new and what is from the template. I don't like it but I don't have a good alternative. I guess we can keep those in as we are reviewing and then delete before merging?

> <!--WHAT DOES THIS ANALYSIS DO AND WHEN WOULD YOU USE IT?-->

Reviewer: I think people are likely to forget to do that last step but maybe I'm wrong. I usually look at the template at the same time when I review.

library(ggplot2)

# We will need this so we can use the pipe: %>%
library(magrittr)
```

## Import data and metadata

Data downloaded from refine.bio include a metadata tab separated values (TSV) file and a data TSV file. 
This chunk of code will read the both TSV files and add them as data frames to your environment.

```{r}
# Read in metadata TSV file
metadata <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your metadata file
  "metadata_SRP078441.tsv" # Replace with the name of your metadata file
))

# Read in data TSV file
df <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your data file
  "SRP078441.tsv" # Replace with the name of your data file
)) %>%
  tibble::column_to_rownames("Gene")
```

Let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>%
  dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(df), metadata$refinebio_accession_code)
```

The information we need to make the comparison is in the `refinebio_title` column of the metadata data.frame.

```{r}
head(metadata$refinebio_title)
```

> ## Set up metadata

Reviewer: I think this section needs more information about the dataset. This could be coming from my 2020 attention span and the fact that I am reviewing this and not _using it_, but repeating yourself a tad here could add welcome context.

We'll use some tidyverse functionality to extract this information into its own column to make it easier to use.

```{r}
metadata <- metadata %>%
  # the last bit of the title, separated by "-" contains the mutation
  # information that we want to extract
  dplyr::mutate(asxl_mutation_status = stringr::word(refinebio_title,
    -1,
    sep = "-"
  )) %>%
  # now let's make this simpler
  dplyr::mutate(asxl_mutation_status = dplyr::case_when(
    grepl("ASXL1|ASXL2", asxl_mutation_status) ~ "asxl_mutation",
    grepl("ASXLwt", asxl_mutation_status) ~ "no_mutation"
  ))
```

Let's take a look at `metadata_df` to see if this worked.

```{r}
# looking at the first 6 rows of the metadata_df and only at the columns that
# contain the title and the mutation status we extracted from the title
head(dplyr::select(metadata, refinebio_title, asxl_mutation_status))
```

Before we set up our model in the next step, we want to check if our modeling variable is set correctly. 
We always want our "control" to to be set as the first level in the variable we provide as our experimental variable. 

```{r}

> str(metadata$asxl_mutation_status)

Reviewer: Why `str()` ? It seems unnecessary to illustrate your point and there is no explanation around it, so I am inclined to take it out.

Author: It is necessary to show them that this is a character and not a factor (what my next sentence after this said). But I didn't explain before doing the step, so I see that that is unclear.

Author: I added words to explain this step some more.

```

Currently, `asxl_mutation_status` is a character.
To make sure it is set how we want, let's mutate it to a factor so we can explicitly set the levels. 

```{r}
# Make asxl_mutation_status a factor and set the levels appropriately
metadata <- metadata %>%
  dplyr::mutate(
    # Here we will set up the factor aspect of our new variable.
    asxl_mutation_status = factor(asxl_mutation_status, levels = c("no_mutation", "asxl_mutation"))
  )
```

Let's double check if the levels are what we want using the `levels()` function. 

```{r}
levels(metadata$asxl_mutation_status)
```

Yes! `no_mutation` is the first level as we want it to be. We're all set and ready to move on to making our `DESeq2Dataset` object. 

## Create a DESeq2Dataset

> We will be using the `DESeq2` package for normalization and differential expression testing, which requires us to format our data into a `DESeqDataSet` object.

Reviewer: ```suggestion
We will be using the `DESeq2` package for differential gene expression analysis, which requires us to format our data into a `DESeqDataSet` object.
```

First we need to prep our gene expression data frame so it's in the format that is compatible with the `DESeqDataSetFromMatrix()` function in the next step. 

```{r}
# We are making our data frame into a matrix and rounding the numbers
gene_matrix <- round(as.matrix(df))
```

Now we need to create `DESeqDataSet` from our expression dataset. 
We use the `asxl_mutation_status` variable we created in the design formula because that will allow us
to model the presence/absence of _ASXL1/2_ mutation.

```{r}
ddset <- DESeqDataSetFromMatrix(
  countData = gene_matrix,
  colData = metadata,
  design = ~asxl_mutation_status
)
```

## Define a minimum counts cutoff

We want to filter out the genes that have not been expressed or that have low expression counts because we want to remove any possible noise from our data before we normalize the data and perform differential expression. 

> We are going to do some pre-filtering to keep only genes with 10 or more reads in total across the samples.

Reviewer: My understanding is that this step, for the analyses that is relevant to this notebook, is most helpful for saving on resources -- e.g., why included these genes as part of the object in memory if you will not be able to (powered to) call them as differentially expressed?

deseq_df <- deseq_results %>%
  # make into data.frame
  as.data.frame() %>%
  # the gene names are rownames -- let's make this it's own column for easy
  # display
  tibble::rownames_to_column("Gene") %>%
  dplyr::mutate(threshold = padj < 0.05) %>%
  # let's sort by statistic -- the highest values should be what is up in the
  # ASX mutated samples
  dplyr::arrange(dplyr::desc(stat))
```

Let's print out what the top results are. 

```{r}
head(deseq_df)
```

### Check results by plotting one gene

To double check what a differentially expressed gene looks like, we can plot one with `DESeq2::plotCounts()` function. 

```{r}
plotCounts(ddset, gene = "ENSG00000196074", intgroup = "asxl_mutation_status")
```
The `mutation` group samples have higher expression of this gene than the control group, which helps assure us that the results are showing us what we are looking for. 

## Save results to TSV 

Write the results table to file.

```{r}
readr::write_tsv(
  deseq_df,
  file.path(
    results_dir,
    "SRP078441_differential_expression_results.tsv" # Replace with a relevant output file name
  )
)
```

## Create a volcano plot

```{r}

> volcano_plot <- ggplot(deseq_df, aes(x = log2FoldChange, y = -log10(padj), color = threshold)) +

Author: We are assuming tidyverse knowledge in general and is part of the "required software" for all modules in the Getting Started section. But I really like the look of `EnhancedVolcano`'s stuff. So I've switched it.

  geom_point() +
  theme_classic() +
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") +
  ggtitle("Volcano plot for mutation status") +
  scale_color_manual(values = c("lightgray", "navy")) +
  theme(legend.position = "none")

# Print out plot here
volcano_plot
```

Save the plot to a PNG. 

```{r}
ggsave(
  plot = volcano_plot,
  file.path(plots_dir, "SRP078441_volcano_plot.png")
) # Replace with a plot name relevant to your data
```

Heatmaps are also a pretty common way to show differential expression results. 
You can take your results from this example and make a heatmap following our heatmap module: TODO: link here. 

# Further learning resources about this analysis

- [`DESeq2` vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)
- [DESeq2 paper](https://doi.org/10.1186/s13059-014-0550-8) [@Love]
- [StatQuest Video: DESeq2, part 1, Library Normalization](https://www.youtube.com/watch?v=UFB993xufUU)

> # Print session info:

Reviewer: I think some of these have not been updated to reflected the new template yet (apologies if that has already been discussed as part of this) review

https://github.com/AlexsLemonade/refinebio-examples/blob/a6ee8c92038972d990d3009f870a18e90ab3819d/template/template_example.Rmd#L217

Author: Yes, I need to take another pass through the template parts since pieces have been updated since when this PR was last reviewed.

You can open this `.Rmd` file in RStudio and follow the rest of these steps from there (See our [section about getting started with R notebooks](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-and-use-rmds) if you are unfamiliar with `.Rmd` files).
Clicking this link will most likely send this to your downloads folder on your computer. 
Move this `.Rmd` file to where you would like this example and its files to be stored.

## Set up your analysis folders 

Good file organization is helpful for keeping your data analysis project on track!
We have set up some code that will automatically set up a folder structure for you. 
Run this next chunk to set up your folders! 

If you have trouble running this chunk, see our [introduction to using `.Rmd`s](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-and-use-rmds) for more resources and explanations. 

```{r}
# Define the file path to the data directory

> data_dir <- file.path("data", "SRP078441") # Replace with path to desired data directory

Author: I've been debating about the order of these steps and what's clearest. (This is a part of the template analysis). But I haven't found  a better order. This is what it is now to summarize:

- Code chunk declares `data_dir` at the same time as the other directories are declared/created. 
- Code chunk creates `data` folder 
- The user drags and drops `SRP078441` folder in the `data` folder. (it's in the instructions)
- Code chunk checks for `data/SRP078441/SRP078441.tsv` file. (which requires the existence already of `data_dir` and the folders it points to)

Fill out the pop up window with your email and our Terms and Conditions:  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/8f1332ce8c5deb0dd93edf6c84b93bae676d9f3b/template/screenshots/download-email.png" width=500>  

<!--DELETE THIS SECTION IF YOU DO WANT QUANTILE NORMALIZED DATA--> 

We are going to use non-quantile normalized data for this analysis.
To get this data, you will need to check the box that says "Skip quantile normalization for RNA-seq samples".
Note that this option will only be available for RNA-seq datasets.

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/b78046845763272d0aa65a077b6a3e46977d5342/template/screenshots/skip-quantile-normalization.png" width=500>

It may take a few minutes for the dataset to process.
You will get an email when it is ready. 

## About the dataset we are using for this example

For this example analysis, we will use this [acute myeloid leukemia dataset](https://www.refine.bio/experiments/SRP078441/rna-seq-of-primary-patient-aml-samples) [@Micol]

> Micol _et._ _al_, performed RNA-seq on primary peripheral blood and bone marrow

Reviewer 2: I'm gonna partly disagree here and suggest it should be `Micol _et al._` with no commas at all. 
 But my real suggestion is to take advantage of Rmarkdown's in text citation, which should handle it all for you!
```suggestion
@Micol performed RNA-seq on primary peripheral blood and bone marrow
```

> samples from AML patients with and without _ASXL1/2_ mutations. 

Author: ? Can you elaborate what you mean here?

> samples from AML patients with and without _ASXL1/2_ mutations. 

Reviewer 3: Just to make it clear for readers that while we can reference <experiment_accession_id> generally in the text, the specific instance of that we are using in this example is SRP078441. I can imagine a reader looking for a directory called experiment_accession_id and not finding it.

> samples from AML patients with and without _ASXL1/2_ mutations. 

Author: Okay. We can replace that too. I think my thought was they can extrapolate this, but I agree with the notion of let's not leave this to chance.

Refine.bio will send you a download button in the email when it is ready. 
Follow the prompt to download a zip file that has a name with a series of letters and numbers and ends in `.zip`. 
Double clicking should unzip this for you and create a folder of the same name.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/8f1332ce8c5deb0dd93edf6c84b93bae676d9f3b/template/screenshots/download-folder-structure.png" width=400> 

For more details on the contents of this folder see [these docs on refine.bio](http://docs.refine.bio/en/latest/main_text.html#rna-seq-sample-compendium-download-folder).  

The `<experiment_accession_id>` folder has the data and metadata TSV files you will need for this example analysis.
Experiment accession ids usually look something like `GSE1235` or `SRP12345`. 

Copy and paste the `<experiment_accession_id>` folder into your newly created `data/` folder.

## Check out our file structure!

Your new analysis folder should contain: 

- The example analysis `.Rmd` you downloaded  
- A folder called "data" which contains:
  - The `<experiment_accession_id>` folder which contains:
    - The gene expression  
    - The metadata TSV  

> 

Author: It shouldn't be empty at this point if they followed the steps in the section `Place the dataset in your new data/ folder` before this.

> 

Reviewer 3: Right... data will have SRP in it, but `plots` and `results` would still be empty right? I think my edit might have obscured my intention. Was thinking you could list `data` `plots` and `results` there since they are all separate folders.

> 

Author: Oh I may have read your comment wrong. :+1:

Your example analysis folder should now look something like this (except with respective experiment accession id and analysis notebook name you are using): 

> <img src="https://github.com/AlexsLemonade/refinebio-examples/raw/b78046845763272d0aa65a077b6a3e46977d5342/template/screenshots/analysis-folder-structure.png" width=400>

Author: It's not inside the `data/` folder but I agree it looks confusing. I'll try rearranging the order and take another screenshot so it doesn't confuse anyone.

> <img src="https://github.com/AlexsLemonade/refinebio-examples/raw/b78046845763272d0aa65a077b6a3e46977d5342/template/screenshots/analysis-folder-structure.png" width=400>

Reviewer 3: ah right -- my bad

> <img src="https://github.com/AlexsLemonade/refinebio-examples/raw/b78046845763272d0aa65a077b6a3e46977d5342/template/screenshots/analysis-folder-structure.png" width=400>

Author: If you were confused, our users will likely be confused, so its a good point. I've switched the screenshot. Hopefully its better.

In order for our example here to run without a hitch, we need these files to be in these locations so we've constructed a test to check before we get started with the analysis. 
Run this chunk to double check that your files are in the right place. 

```{r}
# Check if the gene expression matrix file is in the data directory stored as `data_dir`
file.exists(file.path(data_dir, "SRP078441.tsv"))

# Check if the metadata file is in the data directory stored as `data_dir`
file.exists(file.path(data_dir, "metadata_SRP078441.tsv"))
```

If the chunk above printed out `FALSE` to either of those tests, you won't be able to run this analysis _as is_ until those files are in the appropriate place.

If the concept of a "file path" is unfamiliar to you; we recommend taking a look at our [section about file paths](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#an-important-note-about-file-paths-and-Rmds). 

# Using a different refine.bio dataset with this analysis?

> If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory you created and changing the filenames and paths in the notebook to match these files (we've put comments to signify where you would need to change the code).

Author: Hmm... I do like this idea, but also am worried about people being confused about too much extra comments and hashtags when they are first going through the notebook. I feel like it may be visually distracting or conceptually confusing for particularly new to R users. 

But if you want to show an example of what this could look like, I'm open to seeing if there's a way we can make the "changeme" comments pop. (Could file a separate issue for discussion if you like).

> If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory you created and changing the filenames and paths in the notebook to match these files (we've put comments to signify where you would need to change the code).

Reviewer 3: I like where this is going... could it be achieved by not hardcoding SRPwhatever into those file names but using a variable name to construct the output file paths? The variable name could get defined once at the top of the script and this would be the only place to highlight it/change it?

> If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory you created and changing the filenames and paths in the notebook to match these files (we've put comments to signify where you would need to change the code).

Author: There's more than just file names that folks might need to change when personalizing their analyses, and I'm concerned about having too many variables for them to keep track of. But this is also an interesting idea that we might be able to use in some example analyses.

```{r}
if (!("DESeq2" %in% installed.packages())) {
  # Install this package if it isn't installed yet
  BiocManager::install("DESeq2", update = FALSE)
}
```

Attach the DESeq2:

```{r}
# Attach the DESeq2 library
library(DESeq2)

# Attach the ggplot2 library for plotting
library(ggplot2)

# We will need this so we can use the pipe: %>%
library(magrittr)
```

## Import data and metadata

Data downloaded from refine.bio include a metadata tab separated values (TSV) file and a data TSV file. 
This chunk of code will read the both TSV files and add them as data frames to your environment.

```{r}
# Read in metadata TSV file
metadata <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your metadata file
  "metadata_SRP078441.tsv" # Replace with the name of your metadata file
))

# Read in data TSV file
df <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your data file
  "SRP078441.tsv" # Replace with the name of your data file
)) %>%
  tibble::column_to_rownames("Gene")
```

Let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>%

>   dplyr::select(metadata$refinebio_accession_code)

Author: Can you tell me why you think it would make more sense in that location? I personally like having the "order our samples" step separate so users don't miss the fact that 1) we did this and 2) it's a super important step you should always check for.

>   dplyr::select(metadata$refinebio_accession_code)

Reviewer 3: Totally -- it's really great to have the order check and to be explicit about it. I read the column selection step that precedes the order check as a separate idea that may muddy the waters of the order chunk. I thought of it as first selecting the desired samples and then double checking they are in the right order. As it is, the code achieves this perfectly well!

```{r}
str(metadata$asxl_mutation_status)
```

Currently, `asxl_mutation_status` is a character.
To make sure its set how we want, let's mutate it to a factor so we can explicitly set the levels. 

```{r}
# Make asxl_mutation_status a factor and set the levels appropriately
metadata <- metadata %>%
  dplyr::mutate(
    # Here we will set up the factor aspect of our new variable.
    asxl_mutation_status = factor(asxl_mutation_status, levels = c("no_mutation", "asxl_mutation"))
  )
```

Let's double check if the levels are what we want using the `levels()` function. 

```{r}
levels(metadata$asxl_mutation_status)
```

Yes! `no_mutation` is the first level as we want it to be. We're all set and ready to move on to making our `DESeq2Dataset` object. 

## Create a DESeq2Dataset

We will be using the `DESeq2` package for normalization and differential expression testing, which requires us to format our data into a `DESeqDataSet` object.
First we need to prep our gene expression data frame so it's in the format that is compatible with the `DESeqDataSetFromMatrix()` function in the next step. 

```{r}
df <- df %>%

>   # Bring the "Gene" column back from the rownames

Author: Good point! Removing two lines of code!! 🎉

>   # Bring the "Gene" column back from the rownames

Reviewer 3: Yay! This may impact the sample order checking... ooops beware

  tibble::rownames_to_column("Gene") %>%
  # Mutate numeric variables to be integers
  dplyr::mutate_if(is.numeric, round)
```

Now we need to create `DESeqDataSet` from our expression dataset. 
We use the `asxl_mutation_status` variable we created in the design formula because that will allow us
to model the presence/absence of _ASXL1/2_ mutation.

```{r}
ddset <- DESeqDataSetFromMatrix(
  countData = df,
  tidy = TRUE,
  colData = metadata,
  design = ~asxl_mutation_status
)
```

## Define a minimum counts cutoff

We want to filter out the genes that have not been expressed or that have low expression counts because we want to remove any possible noise from our data before we normalize the data and perform differential expression. 
We are going to do some pre-filtering to keep only genes with 10 or more reads in total across the samples.

```{r}
# Define a minimum counts cutoff and filter `DESeqDataSet` object to include
# only rows that have counts above the cutoff
genes_to_keep <- rowSums(counts(ddset)) >= 10
ddset <- ddset[genes_to_keep, ]
```

## Perform DESeq2 normalization

We are going to use the `rlog()` function from the `DESeq2` package to normalize the data.

> TODO: Add a note/link to section in documentation on why we use `rlog` vs `vst` and `normTransform`.

Author: The TODOs unfortunately have to stay for a bit: https://github.com/AlexsLemonade/refinebio-examples/issues/125#issuecomment-673095903

```{r DESeq}
deseq_object <- DESeq(ddset)
```

Let's take a look at the results table.

```{r deseq_results}
deseq_results <- results(deseq_object)
deseq_results
```

Sort and filter the results. 

```{r arrange-deseq_results}
# this is of class DESeqResults -- we want a data.frame
deseq_df <- deseq_results %>%
  # make into data.frame
  as.data.frame() %>%
  # the gene names are rownames -- let's make this it's own column for easy
  # display
  tibble::rownames_to_column(var = "Gene") %>%
  dplyr::mutate(threshold = padj < 0.05) %>%
  # let's sort by statistic -- the highest values should be what is up in the
  # ASX mutated samples
  dplyr::arrange(dplyr::desc(stat))
```

Let's print out the top results are. 

```{r}
head(deseq_df)
```
### Check results by plotting one gene

To double check what a differentially expressed gene looks like, we can plot one with `DESeq2::plotCounts()` function. 

```{r}
plotCounts(ddset, gene = "ENSG00000196074", intgroup = "asxl_mutation_status")
```
The `mutation` group samples have higher expression of this gene than the control group, which helps assure us that the results are showing us what we are looking for. 

## Save results to TSV 

> Write the `topTable()` output to file.

Author: Whoops. I think this is a remnant from `microarray` where we use `limma`.

```{r  Write results}
readr::write_tsv(
  deseq_df,
  file.path(
    results_dir,

>     "SRP078441_limma_results.tsv" # Replace with a relevant output file name

Author: Oh man! Leftover again (this was adapted from a limma microarrray example). Good catch.

The authors of this data have ASXL mutation status along with other information is stored all in one string (this is not very convenient for us). 
We need to extract the mutation status information into its own column to make it easier to use.

```{r}
metadata <- metadata %>%
  # The last bit of the title, separated by "-" contains the mutation
  # information that we want to extract
  dplyr::mutate(asxl_mutation_status = stringr::word(refinebio_title,
    -1,
    sep = "-"
  )) %>%
  # Now let's summarized the ASXL1 mutation status from this variable
  dplyr::mutate(asxl_mutation_status = dplyr::case_when(
    grepl("ASXL1|ASXL2", asxl_mutation_status) ~ "asxl_mutation",
    grepl("ASXLwt", asxl_mutation_status) ~ "no_mutation"
  ))
```

Let's take a look at `metadata_df` to see if this worked.

```{r}
# looking at the first 6 rows of the metadata_df and only at the columns that
# contain the title and the mutation status we extracted from the title
head(dplyr::select(metadata, refinebio_title, asxl_mutation_status))
```

Before we set up our model in the next step, we want to check if our modeling variable is set correctly. 

> We always want our "control" to to be set as the first level in the variable we provide as our experimental variable. 

Reviewer: ```suggestion
We want our "control" to to be set as the first level in the variable we provide as our experimental variable. 
```

  colData = metadata,
  design = ~asxl_mutation_status
)
```

## Define a minimum counts cutoff

We want to filter out the genes that have not been expressed or that have low expression counts, since these do not have high enough counts to yield reliable differential expression results.
Removing these genes saves on memory usage during the tests. 
We are going to do some pre-filtering to keep only genes with 10 or more reads in total across the samples.

```{r}
# Define a minimum counts cutoff and filter `DESeqDataSet` object to include
# only rows that have counts above the cutoff
genes_to_keep <- rowSums(counts(ddset)) >= 10
ddset <- ddset[genes_to_keep, ]
```

## Run differential expression analysis

We'll use the wrapper function `DESeq()` to do our differential expression analysis.
In our `DESeq2` object we designated our `asxl_mutation_status` variable as the `model` argument. 
Because of this, the `DESeq` function will use groups defined by `asxl_mutation_status` to test for differential expression. 

```{r}
deseq_object <- DESeq(ddset)
```

Let's extract the results table from the `DESeq` object. 

```{r}
deseq_results <- results(deseq_object)
```

> Here we will use `lfcShrink()` function to obtain "shrunk" log fold change estimates based on negative binomial distribution. 

Reviewer: ```suggestion
Here we will use `lfcShrink()` function to obtain "shrunken" log fold change estimates based on negative binomial distribution. 
```
?

Reviewer: I would probably remove the `"` also?

This will add the estimates to your results table. 

> Using `lfcShrink()` can help decrease noise and preserve large differences between groups (it requires that `apeglm` package be installed)

Reviewer: ```suggestion
Using `lfcShrink()` can help decrease noise and preserve large differences between groups (it requires that `apeglm` package be installed).
```

```{r}
deseq_results <- lfcShrink(deseq_object, # This is the original DESeq2 object with DESeq() already having been ran
  coef = 2, # This is based on what log fold change coefficient was used in DESeq(), the default is 2.
  res = deseq_results # This needs to be the DESeq results table
)
```

> Now let's take a peak at what our results table looks like. 

Reviewer: ```suggestion
Now let's take a peek at what our results table looks like. 
```

Author: Lol. 🗻

  # display
  tibble::rownames_to_column("Gene") %>%
  dplyr::mutate(threshold = padj < 0.05) %>%
  # let's sort by statistic -- the highest values should be what is up in the
  # ASXL mutated samples
  dplyr::arrange(dplyr::desc(log2FoldChange))
```

Let's print out what the top results are. 

```{r}
head(deseq_df)
```

### Check results by plotting one gene

To double check what a differentially expressed gene looks like, we can plot one with `DESeq2::plotCounts()` function.

```{r}
plotCounts(ddset, gene = "ENSG00000196074", intgroup = "asxl_mutation_status")
```

The `mutation` group samples have higher expression of this gene than the control group, which helps assure us that the results are showing us what we are looking for. 

## Save results to TSV 

Write the results table to file.

```{r}
readr::write_tsv(
  deseq_df,
  file.path(
    results_dir,
    "SRP078441_differential_expression_results.tsv" # Replace with a relevant output file name
  )
)
```

## Create a volcano plot

We'll use the `EnhancedVolcano` package's main function to plot our data. 

```{r}
volcano_plot <- EnhancedVolcano::EnhancedVolcano(
  deseq_df,
  lab = rownames(deseq_df),
  x = "log2FoldChange",

>   y = "pvalue"

Reviewer: Should this be `padj`? I remember the default cutoff for the plot being smaller than I might expect for a corrected p-value. (It's also going to depend a lot on what you're trying to accomplish with the plot of course.) But it makes sense to me to be _very clear_ about this value in this setting.
